File: submissions/0103-binary-tree-zigzag-level-order-traversal/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function zigzagLevelOrder(root: TreeNode | null): number[][] {
 
    if (!root) {
      return [];
    }
    
    let result = []
    let isReverse = true
    let queue = [root]
    
    while (queue.length) {
        let currentLevelLength = queue.length;
         let newQueue = [];
        
        for (let i = 0; i < currentLevelLength; i++){
            let node = queue.shift();
            newQueue.push(node.val)
           
            if(node.right){
                queue.push(node.right)
            }
            if(node.left){
                queue.push(node.left)
                
            }
           
        }
     
         if (isReverse) {
            newQueue.reverse();
        }
            
            result.push(newQueue)
        isReverse = !isReverse
    }
    return result;
};

File: submissions/0103-binary-tree-zigzag-level-order-traversal/solution.py
from collections import deque
from typing import List, Optional

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        queue = deque([root])
        ans = []
        level = 0
        
        while queue:
            curr_len = len(queue)
            nodes = []
            
            for _ in range(curr_len):
                node = queue.popleft()
                nodes.append(node.val)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            if level % 2 == 1:
                nodes.reverse()
            
            ans.append(nodes)
            level += 1
        
        return ans

